### PR TITLE
GMC Agent + Binary metric exporter.

### PR DESCRIPTION
**Pull request by smoldachev**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommands.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+/**
+ * Management cluster command.
+ */
+public class ManagementCommands implements Command<ManagementArguments> {
+    /** */
+    private ManagementArguments args;
+
+    /** {@inheritDoc} */
+    @Override public void printUsage(Logger log) {
+        Command.usage(log, "Enable management:", MANAGEMENT, ManagementCommandList.ENABLE.text());
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommands.java]: Lets use static import for `Command.usage(xxx)`;

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandList.java [hunk 2]
```diff
@@ -62,7 +63,10 @@
     READ_ONLY_ENABLE("--read-only-on", new ClusterReadOnlyModeEnableCommand()),
 
     /** */
-    READ_ONLY_DISABLE("--read-only-off", new ClusterReadOnlyModeDisableCommand());
+    READ_ONLY_DISABLE("--read-only-off", new ClusterReadOnlyModeDisableCommand()),
+
+    /** */
+    MANAGEMENT("--management", new ManagementCommands());
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandList.java]: @agoncharuk I think we need to hide this new command same as we do for 'rolling upgrade and baseline autoadjustment'. Agree?

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommandList.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.commandline.management;
+
+import org.apache.ignite.internal.commandline.argument.CommandArg;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Set of management commands.
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommandList.java]: Lets describe in javadocs command syntax and usages.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java [hunk 4]
```diff
@@ -1944,8 +1944,9 @@ public void sendToGridTopic(
         Message msg,
         byte plc
     ) throws IgniteCheckedException {
-        assert F.find(nodes, null, F.localNode(locNodeId)) == null :
-            "Internal Ignite code should never call the method with local node in a node list.";
+// TODO GG-21721 temporary allow this method on local node.
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/communication/GridIoManager.java]: @agoncharuk  I think we should fix GG-21721 before this PR will be merged to master!!!

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java [hunk 5]
```diff
@@ -203,6 +202,8 @@ public ClusterMetricsImpl(GridKernalContext ctx, long nodeStartTime) {
         this.ctx = ctx;
         this.nodeStartTime = nodeStartTime;
 
+        //MetricRegistry mreg = new MetricRegistry(SYS_METRICS, SYS_METRICS, ctx.log(getClass()));
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/ClusterMetricsImpl.java]: @agoncharuk Why this code is commented? Without TODO?

File: modules/core/src/main/java/org/apache/ignite/internal/metric/IoStatisticsHolderCache.java [hunk 6]
```diff
@@ -53,7 +54,7 @@
      * @param grpId Group id.
      * @param mmgr Metric manager.
      */
-    public IoStatisticsHolderCache(String cacheName, int grpId, GridMetricManager mmgr) {
+    public IoStatisticsHolderCache(String cacheName, int grpId, GridMetricManager mmgr, IgniteLogger log) {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/metric/IoStatisticsHolderCache.java]: @agoncharuk Looks like `log` argument is not used?

File: modules/core/src/main/java/org/apache/ignite/internal/metric/IoStatisticsHolderIndex.java [hunk 7]
```diff
@@ -72,7 +73,9 @@ public IoStatisticsHolderIndex(
         IoStatisticsType type,
         String cacheName,
         String idxName,
-        GridMetricManager mmgr) {
+        GridMetricManager mmgr,
+        IgniteLogger log
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/metric/IoStatisticsHolderIndex.java]: @agoncharuk  Not used `log`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/DataRegionMetricsImpl.java [hunk 8]
```diff
@@ -114,8 +115,10 @@
      * @param dataRegionMetricsProvider Data region metrics provider.
      */
     public DataRegionMetricsImpl(DataRegionConfiguration memPlcCfg,
-        GridMetricManager mmgr,
-        DataRegionMetricsProvider dataRegionMetricsProvider) {
+                                 GridMetricManager mmgr,
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/DataRegionMetricsImpl.java]: 1. Not used log.
2. Looks like arguments formatted not by Ignite / GG code style.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java [hunk 9]
```diff
@@ -180,6 +184,22 @@
             DFLT_LONG_OPERATIONS_DUMP_TIMEOUT
     );
 
+    /** Metric name for total system time on node. */
+    public static final String METRIC_TOTAL_SYSTEM_TIME = "totalNodeSystemTime";
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java]: @agoncharuk Looks like we are implementing metrics here, but I thought this issue is about binary protocol. It is a bad practice to solve several problems in one issue. Lets move this changes to separate issue?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/gmc/ManagementConsoleProcessor.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public ManagementConfiguration configuration() {
+        return cfg;
+    }
+
+    /**
+     * @return No-op processor usage exception;
+     */
+    private IgniteException processorException() {
+        return new IgniteException("Current Ignite configuration does not support Management Console.");
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/gmc/ManagementConsoleProcessor.java]: @agoncharuk I think we should use anoth message here. May be `Cluster has no management agent `?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/gmc/package-info.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * <!-- Package description. -->
+ * Contains <b>GridGain Managment Console</b> functionality.
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/gmc/package-info.java]: @agoncharuk are we OK with "GridGain" here?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/job/GridJobProcessor.java [hunk 12]
```diff
@@ -301,9 +301,11 @@ public GridJobProcessor(GridKernalContext ctx) {
         jobExecLsnr = new JobExecutionListener();
         discoLsnr = new JobDiscoveryListener();
 
-        cpuLoadMetric = ctx.metric().registry(SYS_METRICS).findMetric(CPU_LOAD);
+        cpuLoadMetric = (DoubleMetric)ctx.metric().registry(SYS_METRICS).findMetric(CPU_LOAD);
 
-        MetricRegistry mreg = ctx.metric().registry(JOBS_METRICS);
+        //MetricRegistry mreg = new MetricRegistry(JOBS_METRICS, JOBS_METRICS, log);
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/job/GridJobProcessor.java]: @agoncharuk Why this line was commented without TODO?

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommandList.java (hunk 3 above)
Comment (nva) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/management/ManagementCommandList.java]: @akuznetsov-gridgain No, we did this in visor cmd and it's totally uncomfortable when you modify command you need change: help command output, javadoc. Also other commands doesn't do this

**Comment by smoldachev** [comment]
Close this PR look at https://github.com/gridgain/gridgain/pull/640

**Pull request closed without merging**